beacon-chain/sync/initial-sync/blocks_fetcher.go

```diff
@@ -640,70 +598,557 @@ func (f *blocksFetcher) fetchBlobsFromPeer(ctx context.Context, bwb []blocks2.Bl
+		return false, 0, 0, nil
 	}
 
-	colIdxs := make([]uint64, 0, len(custodyCols))
-	for c := range custodyCols {
-		colIdxs = append(colIdxs, c)
+	// Find the first block which contains blob commitments.
+	firstBlockWithCommitmentIndex := -1
+	for i := firstWindowIndex; i < len(bwb); i++ {
+		// Is there any blob commitment in this block?
+		commits, err := bwb[i].Block.Block().Body().BlobKzgCommitments()
+		if err != nil {
+			return false, 0, 0, errors.Wrap(err, "blob KZG commitments")
+		}
+
+		if len(commits) > 0 {
+			firstBlockWithCommitmentIndex = i
+			break
+		}
 	}
-	req.Columns = colIdxs
-	peers = f.filterPeers(ctx, peers, peersPercentagePerRequest)
-	// We dial the initial peer first to ensure that we get the desired set of columns.
-	wantedPeers := append([]peer.ID{pid}, peers...)
-	bestPeers := f.hasSufficientBandwidth(wantedPeers, req.Count)
-	// We append the best peers to the front so that higher capacity
-	// peers are dialed first. If all of them fail, we fallback to the
-	// initial peer we wanted to request blobs from.
-	peers = append(bestPeers, pid)
-	for i := 0; i < len(peers); i++ {
-		p := peers[i]
-		nid, err := p2p.ConvertPeerIDToNodeID(pid)
+
+	if firstBlockWithCommitmentIndex == -1 {
+		// There is no block with blob commitments.
+		return false, 0, 0, nil
+	}
+
+	// Find the last block which contains blob commitments.
+	lastBlockWithCommitmentIndex := len(bwb) - 1
+	for i := lastBlockWithCommitmentIndex; i >= firstBlockWithCommitmentIndex; i-- {
+		// Is there any blob commitment in this block?
+		commits, err := bwb[i].Block.Block().Body().BlobKzgCommitments()
 		if err != nil {
-			return nil, err
+			return false, 0, 0, errors.Wrap(err, "blob KZG commitments")
+		}
+
+		if len(commits) > 0 {
+			lastBlockWithCommitmentIndex = i
+			break
+		}
+	}
+
+	return true, firstBlockWithCommitmentIndex, lastBlockWithCommitmentIndex, nil
+}
+
+// blocksWithMissingDataColumnsBoundaries finds the first and last block in `bwb` that contain at least one missing data column.
+func (f *blocksFetcher) blocksWithMissingDataColumnsBoundaries(
+	bwb []blocks2.BlockWithROBlobs,
+	firstIndex, lastIndex int,
+	localCustodyColumns map[uint64]bool,
+) (bool, int, int) {
+	// Find the first block for which some data columns to retrieve are not in our store.
+	firstBlockWithMissingColumnsIndex := -1
+	for i := firstIndex; i < len(bwb); i++ {
+		root := bwb[i].Block.Root()
+		allColumnsAreAvailable := f.bs.Summary(root).AllDataColumnsAvailable(localCustodyColumns)
+
+		if !allColumnsAreAvailable {
+			firstBlockWithMissingColumnsIndex = i
+			break
 		}
+	}
+
+	if firstBlockWithMissingColumnsIndex == -1 {
+		// There is no block with missing data columns.
+		return false, 0, 0
+	}
+
+	// Find the last block for which some data columns to retrieve are not in our store.
+	lastBlockWithMissingColumnsIndex := lastIndex
+	for i := lastIndex; i >= firstBlockWithMissingColumnsIndex; i-- {
```

Comment:
I do not understand this loop. the moment it is negatively incremented, the condition would be false as `i < firstBlockWithMissingColumnsIndex` . Do we meant to only check this once ?

Reply:
The loop starts at `lastBlockWithMissingColumnsIndex` and goes backward until:
- `firstBlockWithMissingColumnsIndex` is met, or if
- A block which at least one data columns we should custody and which we actually don't custody is met.